### PR TITLE
Bug 1822200: Custom dns name support

### PR DESCRIPTION
**Pull request by michaelgugino**
This PR adds support for custom domains provided via DHCP options, primarily for the usecase of private VPCs (but possibly other usecases as well).  These domains are necessary to be present in the node addresses to approve CSR requests.

Support space delimited custom domains

add unit tests

### CODE REVIEW

File: pkg/actuators/machine/reconciler.go [hunk 1]
```diff
@@ -187,7 +187,30 @@ func (r *Reconciler) update() error {
...
-	r.machineScope.setProviderStatus(newestInstance, conditionSuccess())
+	vpc, err := r.awsClient.DescribeVpcs(&ec2.DescribeVpcsInput{
+		VpcIds: []*string{newestInstance.VpcId},
+	})
+	if err != nil {
+		klog.Errorf("%s: error describing vpc: %v", r.machine.Name, err)
+		return err
+	}
+
+	dhcp, err := r.awsClient.DescribeDHCPOptions(&ec2.DescribeDhcpOptionsInput{
```
Comment (michaelgugino) [inline pkg/actuators/machine/reconciler.go:198]: Is this information able to change over the life of the VPC?  If not, maybe we could stuff the domainName in machine's status and only retrieve this if it's missing.

Comment (JoelSpeed) [reply pkg/actuators/machine/reconciler.go:198]: DNS Settings can be changed at any time in the VPC lifecycle, so I think it makes sense for this to be done on `update` to check if they've changed (docs: https://docs.aws.amazon.com/vpc/latest/userguide/vpc-dns.html#vpc-dns-updating)

File: pkg/actuators/machine/reconciler.go [hunk 2]
```diff
@@ -187,7 +187,30 @@ func (r *Reconciler) update() error {
...
+		DhcpOptionsIds: []*string{vpc.Vpcs[0].DhcpOptionsId},
+	})
+	if err != nil {
+		klog.Errorf("%s: error describing dhcp: %v", r.machine.Name, err)
+		return err
+	}
+	var domainName string
+	for _, i := range dhcp.DhcpOptions[0].DhcpConfigurations {
+		if *i.Key == "domain-name" {
+			klog.Infof("Domain value %v", i.Values)
```
Comment (JoelSpeed) [inline pkg/actuators/machine/reconciler.go:208]: Do we want this to be logged? Is this debug?

File: pkg/actuators/machine/reconciler.go [hunk 3]
```diff
@@ -187,7 +187,30 @@ func (r *Reconciler) update() error {
...
+			domainName = i.Values[0].String()
```
Comment (JoelSpeed) [inline pkg/actuators/machine/reconciler.go:209]: Should we break once we've found the value?

File: pkg/actuators/machine/reconciler.go [hunk 4]
```diff
@@ -187,7 +187,30 @@ func (r *Reconciler) update() error {
...
+
+	dhcp, err := r.awsClient.DescribeDHCPOptions(&ec2.DescribeDhcpOptionsInput{
+		DhcpOptionsIds: []*string{vpc.Vpcs[0].DhcpOptionsId},
+	})
+	if err != nil {
+		klog.Errorf("%s: error describing dhcp: %v", r.machine.Name, err)
+		return err
+	}
+	var domainName string
+	for _, i := range dhcp.DhcpOptions[0].DhcpConfigurations {
```
Comment (JoelSpeed) [inline pkg/actuators/machine/reconciler.go:206]: How hard would it be to add a test that covers this within the reconciler tests?

File: pkg/actuators/machine/reconciler.go (hunk 1 above)
Comment (michaelgugino) [reply pkg/actuators/machine/reconciler.go:198]: Yeah, probably is the safest solution.  Just feels like more API quota exhaustion.  If you have lots of machines, that's two extra API calls per reconcile that could really add up.

File: pkg/actuators/machine/reconciler.go (hunk 2 above)
Comment (michaelgugino) [reply pkg/actuators/machine/reconciler.go:208]: I meant to remove that, but I'm going to leave that in for now to confirm what AWS actually spits out.  I'll remove it in a few.

File: pkg/actuators/machine/reconciler.go (hunk 3 above)
Comment (michaelgugino) [reply pkg/actuators/machine/reconciler.go:209]: Yeah, I meant to stick that in there.

File: pkg/actuators/machine/reconciler.go (hunk 4 above)
Comment (michaelgugino) [reply pkg/actuators/machine/reconciler.go:206]: It would be fairly non trivial.  I would prefer to refactor update() from a method to a function in order to make testing here easier.  We're over using methods and it makes mocking tests much more burdensome.

So, I don't think the juice is worth the squeeze here.

Comment (enxebre) [inline pkg/actuators/machine/reconciler.go:206]: can `DhcpOptions` ever come empty making [0] to panic?

File: pkg/actuators/machine/reconciler.go [hunk 5]
```diff
@@ -187,7 +187,30 @@ func (r *Reconciler) update() error {
...
+		if *i.Key == "domain-name" {
```
Comment (enxebre) [inline pkg/actuators/machine/reconciler.go:207]: may be put `"domain-name"` in a constant

File: pkg/actuators/machine/reconciler.go [hunk 6]
```diff
@@ -187,7 +187,30 @@ func (r *Reconciler) update() error {
 
 	klog.Infof("Updated machine %s", r.machine.Name)
 
-	r.machineScope.setProviderStatus(newestInstance, conditionSuccess())
+	vpc, err := r.awsClient.DescribeVpcs(&ec2.DescribeVpcsInput{
```
Comment (enxebre) [inline pkg/actuators/machine/reconciler.go:190]: can we wrap all this up into a function e.g `getCustomDNS` that we can unit test?

File: pkg/actuators/machine/reconciler.go (hunk 3 above)
Comment (enxebre) [reply pkg/actuators/machine/reconciler.go:209]: Could Values[0] ever panic, do we need to sanity check Values len?

**Comment by enxebre** [comment]
Can we please add a description to this PR and squash commits?

File: pkg/actuators/machine/reconciler.go [hunk 7]
```diff
@@ -187,11 +188,48 @@ func (r *Reconciler) update() error {
...
+		VpcIds: []*string{vpcID},
+	})
+	if err != nil {
+		klog.Errorf("%s: error describing vpc: %v", r.machine.Name, err)
+		return nil, err
+	}
+
+	dhcp, err := r.awsClient.DescribeDHCPOptions(&ec2.DescribeDhcpOptionsInput{
+		DhcpOptionsIds: []*string{vpc.Vpcs[0].DhcpOptionsId},
```
Comment (enxebre) [inline pkg/actuators/machine/reconciler.go:212]: wouldn' thist panic if the aws backend ever return a vpc.Vpcs[0] nil?

File: pkg/actuators/machine/utils.go [hunk 8]
```diff
@@ -304,6 +304,10 @@ func extractNodeAddresses(instance *ec2.Instance) ([]corev1.NodeAddress, error)
 	if privateDNSName != "" {
 		addresses = append(addresses, corev1.NodeAddress{Type: corev1.NodeInternalDNS, Address: privateDNSName})
 		addresses = append(addresses, corev1.NodeAddress{Type: corev1.NodeHostName, Address: privateDNSName})
+		for _, dN := range domainNames {
```
Comment (enxebre) [inline pkg/actuators/machine/utils.go:307]: nit: camelcase here `dN` seems odd, should probably just name it dn

File: pkg/actuators/machine/utils.go [hunk 9]
```diff
@@ -304,6 +304,10 @@ func extractNodeAddresses(instance *ec2.Instance) ([]corev1.NodeAddress, error)
...
+			customHostName := strings.Join([]string{strings.Split(privateDNSName, ".")[0], dN}, ".")
+			addresses = append(addresses, corev1.NodeAddress{Type: corev1.NodeHostName, Address: customHostName})
```
Comment (enxebre) [inline pkg/actuators/machine/utils.go:309]: we should be setting also `NodeInternalDNS`

File: pkg/actuators/machine/reconciler.go [hunk 10]
```diff
@@ -187,11 +188,48 @@ func (r *Reconciler) update() error {
 
 	klog.Infof("Updated machine %s", r.machine.Name)
 
-	r.machineScope.setProviderStatus(newestInstance, conditionSuccess())
+	domainNames, err := r.getCustomDomainFromDHCP(newestInstance.VpcId)
```
Comment (enxebre) [inline pkg/actuators/machine/reconciler.go:191]: I don't thins this should be a parameter for `setProviderStatus`. I think this should be called from within `extractNodeAddresses` or `setProviderStatus`. i.e they should never run with out this.

File: pkg/actuators/machine/reconciler.go [hunk 11]
```diff
@@ -187,11 +188,48 @@ func (r *Reconciler) update() error {
...
+	if err != nil {
+		return err
+	}
+
+	r.machineScope.setProviderStatus(newestInstance, conditionSuccess(), domainNames)
 
 	return r.requeueIfInstancePending(newestInstance)
 }
 
+func (r *Reconciler) getCustomDomainFromDHCP(vpcID *string) ([]string, error) {
```
Comment (enxebre) [inline pkg/actuators/machine/reconciler.go:202]: can we at minimum have a unit for this function? e.g similar to `TestGetBlockDeviceMappings`

File: pkg/actuators/machine/machine_scope_test.go [hunk 12]
```diff
@@ -286,3 +290,55 @@ func TestPatchMachine(t *testing.T) {
...
+	testCases := []struct {
+		description string
+		expected    []string
+	}{
+		{
+			description: "Returns two",
+			expected:    []string{"openshift.com", "openshift.io"},
+		},
+	}
+
```
Comment (elmiko) [inline pkg/actuators/machine/machine_scope_test.go:331]: it seems like there are several conditions where `getCustomDomainFromDHCP` can return `nil, nil`, does it make sense to add those as unit tests as well?

File: pkg/actuators/machine/machine_scope.go [hunk 13]
```diff
@@ -18,6 +19,9 @@ const (
 	userDataSecretKey = "userData"
 )
 
+// dhcpDomainKeyName is a variable so we can reference it in unit tests.
+var dhcpDomainKeyName = "domain-name"
```
[the remark this answers is not in the document]
Comment (michaelgugino) [reply pkg/actuators/machine/machine_scope.go:23]: @elmiko need to take a reference in unit tests.  Can't take a reference to a constant.

File: pkg/actuators/machine/machine_scope_test.go (hunk 12 above)
Comment (michaelgugino) [reply pkg/actuators/machine/machine_scope_test.go:331]: Not really, IMO.

Comment (enxebre) [reply pkg/actuators/machine/machine_scope_test.go:331]: can you elaborate why not? as far as I can see it definitely make sense and has value to cover more code paths, e.g description: "Returns one", description: "Returns empty"?

File: pkg/actuators/machine/utils_test.go [hunk 14]
```diff
@@ -54,6 +56,32 @@ func TestExtractNodeAddresses(t *testing.T) {
...
+						PrivateIpAddresses: []*ec2.InstancePrivateIpAddress{
+							{
+								Primary:          aws.Bool(true),
+								PrivateIpAddress: aws.String("10.0.0.5"),
+							},
+						},
+					},
+				},
+			},
+			expectedAddresses: []corev1.NodeAddress{
```
Comment (enxebre) [inline pkg/actuators/machine/utils_test.go:77]: It'd be nice to include a case that checks we don't duplicate as per https://github.com/openshift/cluster-api-provider-aws/pull/318/files#diff-9d8c56b0090958d82aa8280b211d4bf8R309-R311

**Comment by enxebre** [comment]
lgtm other than https://github.com/openshift/cluster-api-provider-aws/pull/318#discussion_r421416357 and https://github.com/openshift/cluster-api-provider-aws/pull/318#discussion_r421417888

**Comment by enxebre** [comment]
/hold
to make sure it does not get in before squashing.

File: pkg/actuators/machine/machine_scope_test.go (hunk 12 above)
Comment (michaelgugino) [reply pkg/actuators/machine/machine_scope_test.go:331]: I'm not a huge fan of mocking up fake cloud responses that don't materially drive decision making in our code.  If the cloud sends back some empty stuff, cool, we don't have any domains.  This case (returning nil,nil) is actually covered by several other tests because we don't mock any return from DescribeVPCS or DescribeDhcpOptions.

The only case where I can see something actually being empty is vpc.Vpcs[0].DhcpOptionsId.  It's maybe possible to have a VPC without a DHCP option set assigned (and this is the case that is covered via other unit tests).  It's also probably possible to have a DHCP option set assigned without any actual options, and possibly one defined with some options but nothing for domain-name, but does that actually happen ever?  Probably not, or at least not in a VPC you expect to work.  And even in that very limited case, the absence doesn't affect us.  We're not going to fail on missing info.

Since we're only driving this function with a single input (vpcID) which is entirely contrived in testing, adding a bunch of statements to simulate the AWS API seems pointless.  Firstly, there's no guarantee we're going to mock all the possible responses, let alone mock them correctly.  Secondly, we're validating the output of this function as a valid input to another function (extractNodeAddresses) via other tests, albeit the nil,nil case.  We're validating that this function will return a list of strings.  We validate in extractNodeAddresses that a list of strings does what we want.

The more important tests would be validating the error conditions, because that will actually break existing users.  But again, that's just mocking an established library and API, so there's no value in doing that.  If this code breaks, it will 'fail open' unless we segfault, which we should be defended against properly now.

File: pkg/actuators/machine/utils_test.go (hunk 14 above)
Comment (michaelgugino) [reply pkg/actuators/machine/utils_test.go:77]: Done.

File: pkg/actuators/machine/machine_scope_test.go (hunk 12 above)
Comment (elmiko) [reply pkg/actuators/machine/machine_scope_test.go:331]: that makes sense to me. my main concern with the original question was just to make sure we are exercising all the possible return paths in that function. i wouldn't expect a full mock of the upstream api, just enough to cover the various return paths. but i get what you are saying about the varied paths that could lead to those error conditions based on the range of possible inputs.

Comment (enxebre) [reply pkg/actuators/machine/machine_scope_test.go:331]: >If the cloud sends back some empty stuff, cool, we don't have any domains...

This func has 5 return calls and we only certainly know it does what it's meant for one of single return path and only in the case we expect "Returns two".

We don't know what the func does if DhcpOptions legitimately has only 1 domain or has no domains. And so we don't know if any future change to the implementation of the func would break that behaviour.

To me seems a lot of value we could get for the price of adding a few lines including 2 more testCases.

Any ways not a blocker to me...

**Comment by enxebre** [comment]
@michaelgugino any reason to split in two commits? feel free to completely smash my initial dev commit, we don't need that in history
